### ISSUE TITLE
Consider ARM64EC support as part of the 64 bit migration

### ISSUE THREAD

**Issue by LeonarddeR** [nvaccess/nvda#18445]
### Is your feature request related to a problem? Please describe.
When migrating to 64-bit NVDA, we'd still be using an AMD64 version on ARM64 systems. A fully native arm64 version of NVDA would require a separate distribution of NVDA, since it needs a separate interpreter. I assume that is out of scope.

### Describe the solution you'd like
Build all NVDA helper code for ARM64EC as well, this means, we'd have 4 lib folders in the end:
- X86, containing NVDA Helper remote and vbuf backends for x86 compatibility
- ARM64, containing NVDA Helper remote and vbuf backends for compatibility with native ARM64 processes
- X64, containing all the libraries, helper local, remote, liblouis, etc.
- ARM64ec: also contains all the libraries like x64, but loaded on ARM64 systems, since the AMD64 python bundled with NVDA should be able to load ARM64EC code within the AMD64 process.

Basically, the native arch for all dll's loaded into NVDA should be X64 on AMD64 and ARM64ec on arm64.

### Describe alternatives you've considered
Just run native AMD64 code on ARM64, but that's probably slower then the ARM64EC code, which is optimized for ARM.

### Additional context
Related to #18207 

[ARM64EC](https://learn.microsoft.com/en-us/windows/arm/arm64ec) is a special ABI from Microsoft  that allows binaries to run natively on ARM64 hardware while remaining binary-compatible with x64. This means ARM64EC code can directly interoperate with x64 processes. On Windows ARM systems, running x64 code normally requires full emulation, which is slower. By using ARM64EC, we can compile performance-critical parts of NVDA to run natively on ARM64 while still relying on existing x64 code (e.g. Python interpreter) for the rest.

According to Microsoft, ARM64EC has native performance on ARM64 systems, compared to limited performance for emulated X64 code.

What this means for NVDA on ARM64:

- No longer use x64 NVDAHelper, but compile NVDAHelper for ARM64EC instead
- Also include `nvdaHelperLocal.dll`, `nvdaHelperLocalWin10.dll` and `UIARemote.dll`.
- There will still be three architectures we need to support on ARM64. x86 is covered by the x86 libraries, arm64 is covered by the arm64 libraries, x64 is covered by the arm64ec libraries.

**Comment by LeonarddeR** [comment]
Several additional details:

* For OSARA, https://github.com/jcsteh/osara/pull/1289 creates an ARM64EC build
* https://github.com/SCons/scons/discussions/4745 was created to have first class support in SCons
* I investigated building an [ARM64X](https://learn.microsoft.com/en-us/windows/arm/arm64x-pe) library to see how injection behaves with binaries that contain both ARM64EC and native ARM64 code. In the most ideal situation, such a binary could inject into both arm64 and x64 processes on ARM. That would have meant we'd need only one  remote loader on arm64. I created a proof of concept in https://github.com/LeonarddeR/scons_arm64_poc, based on NVDAHelper focus hooking, but as I already thought, it uses the architecture of the host binary.

**Comment by SaschaCowley** [comment]
@LeonarddeR we're having trouble following this proposal. Could you please restate your proposal and why it's necessary? Please include clear user stories.

**Comment by LeonarddeR** [comment]
@SaschaCowley I'm sorry for the vagueness. I added additional context to the initial comment. I hope that's enough.
I have an ARm64 VM on Azure to test with, so if you want to see real performance comparisons, I can probably provide them.

**Comment by mzanm** [comment]
hello. Could Espeak and Sonic be also considered with an ARM64EC variant? Thanks.

**Comment by gexgd0419** [comment]
So here's a compatibility table that shows what kind of DLL can be loaded by what kind of process:

| DLL arch | x86 process | x64 process | Arm64EC process | Arm64 process |
| :-: | :-: | :-: | :-: | :-: |
| x86 DLL | Yes | No | No | No |
| x64 DLL | No | Yes | Yes | No |
| Arm64EC DLL | No | Yes | Yes | No |
| Arm64 DLL | No | No | No | Yes |
| Arm64X DLL | No | Yes | Yes | Yes |

Only Arm64X DLLs can be loaded by all kinds of 64-bit processes on Windows on Arm.

An Arm64X DLL is a combination of Arm64 code and Arm64EC code, and some shared resources. When loaded by an x64/Arm64EC process, it behaves like an Arm64EC DLL; and when loaded by an Arm64 process, it behaves like an Arm64 DLL. For systems that don't know about Arm64X, such DLLs behave like Arm64 DLLs, so it cannot be used on x86/x64 platforms.

> * I investigated building an [ARM64X](https://learn.microsoft.com/en-us/windows/arm/arm64x-pe) library to see how injection behaves with binaries that contain both ARM64EC and native ARM64 code. In the most ideal situation, such a binary could inject into both arm64 and x64 processes on ARM. That would have meant we'd need only one  remote loader on arm64. I created a proof of concept in https://github.com/LeonarddeR/scons_arm64_poc, based on NVDAHelper focus hooking, but as I already thought, it uses the architecture of the host binary.

I guess that's because when the host process loads the Arm64X DLL, the DLL will be treated as either an Arm64EC DLL or an Arm64 DLL. Then, when the host process calls `SetWinEventHook`, only that "part" of the Arm64X DLL will be injected. But you can still use different host processes to load the same Arm64X DLL both as Arm64EC and as Arm64.

This is for Windows hooks though. If you inject DLLs by creating remote threads and let the process load the DLL itself, then yes, Arm64X DLLs can be injected into both x64/Arm64EC and Arm64 processes.

**Comment by gexgd0419** [comment]
Some backgrounds:

We already know that on an x64 system, x86 and x64 programs can coexist. But x86 and x64 are different architectures, so you cannot load an x64 DLL in an x86 process, or vice versa. An x64 processor can execute both x86 code and x64 code natively, so x86 programs can run on an x64 system without much performance loss.

Then there's Windows on ARM, where the CPU has another architecture: ARM64. ARM CPUs cannot execute x86 or x64 code, only ARM code. However, most Windows programs are still x86/x64 only, so Microsoft introduces an x86/x64 emulator, which allows Windows on ARM to run x86 and x64 apps with emulation. But emulation can cause performance loss. So it's recommended to use a native ARM64 version of software if there is one.

For developers though, migrating to ARM64 might not be that easy. You may have some parts that don't have an ARM64 version, and/or you cannot compile an ARM64 version of them. But as ARM64 and x86/x64 are different architectures, you cannot load an x86/x64 library into your ARM64 process, forcing you to either separate them into another process, or drop the ARM64 support entirely.

To make this migration easier, a new architecture (or binary interface) called Arm64EC or "Arm64 emulation compatible" is introduced. Arm64EC is like Arm64 but in x64's clothing. Because Arm64EC still appears to be x64, x64 code can load Arm64EC libraries, call Arm64EC code, or even hook Arm64EC functions, seamlessly. Arm64EC code can also call x64 code or other Arm64EC code. You can just treat Arm64EC as if it's x64 most of the time. But other than the x64's clothing, Arm64EC still has ARM64 code inside that can be natively executed by ARM CPUs, which is more efficient than pure x64 code that requires emulation.

Arm64EC sacrifices the compatibility with pure Arm64 code for the compatibility with x64 code. So you still cannot load Arm64 libraries into Arm64EC processes, or vice versa. You still have to choose between x64 and Arm64 in some cases, as some libraries do not have an Arm64EC version.

What also gets sacrificed is some performance. Compared to pure Arm64 code, Arm64EC has some performance penalties. But it's still better than pure x64 code though.

**Comment by josephsl** [comment]
Hi, there is also a compatible issue to consider which relates to x64 emulation on ARM64: once we transition to x64 Python runtime, we will lose support for Windows 10 on ARM which doesn’t have x64 emulation. Thanks.

**Comment by gexgd0419** [comment]
The first Windows version to support ARM64 was Windows 10 1709, which only supports 32-bit x86 emulation, not x64 emulation. x64 emulation is only supported on Windows 11, not Windows 10.

Arm64EC is also not supported on Windows 10. Arm64X is supported, although Windows 10 will treat Arm64X DLLs as Arm64 DLLs.

So if we want to keep the compatibility with Windows 10 on ARM, we would have to either keep an x86 version, or make a pure ARM64 version. The incompatibility between ARM64 and x64 means that a cross-process bridge for x64 drivers may be needed on ARM64 builds.

**Comment by gexgd0419** [comment]
Windows 10 will reach end of support on October 14, 2025. So maybe we can just drop support for Windows 10 on ARM.

**Comment by LeonarddeR** [comment]
We should certainly drop support for Windows 10 in my humble opinion, if only because Windows 10 on Arm users hardly have a valid reason to stay on Windows 10.

**Comment by LeonarddeR** [comment]
@SaschaCowley Do you still need more info or do you think the description is clear enough now?
I'm happy to prototype something against the `wip-buildX64` branch if you like.

**Comment by SaschaCowley** [comment]
I think we need to discuss this more internally, however you can feel free to prototype it if you like. Certainly if you can prototype it and demonstrate that it is not a significant amount of work to maintain and/or provides demonstrable performance improvements, that would improve the likelihood of us accepting the proposal.

**Comment by LeonarddeR** [comment]
> Certainly if you can prototype it and demonstrate that it is not a significant amount of work to maintain and/or provides demonstrable performance improvements, that would improve the likelihood of us accepting the proposal.

No, it doesn't take that much effort really. I will file a pr soon.

**Comment by LeonarddeR** [comment]
I let ChatGPT wrote a quick performance test app:

```CPP
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <random>

using namespace std;
using namespace chrono;

bool is_prime(int n) {
    if (n <= 1) return false;
    for (int i = 2; i <= sqrt(n); ++i)
        if (n % i == 0)
            return false;
    return true;
}

void calculate_primes(int limit) {
    auto start = high_resolution_clock::now();

    int count = 0;
    for (int i = 2; i <= limit; ++i)
        if (is_prime(i)) ++count;

    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<seconds>(end - start).count();
    cout << "Total primes up to " << limit << ": " << count << " (took " << elapsed << " seconds)" << endl;
}

void matrix_multiply(int size) {
    auto start = high_resolution_clock::now();

    vector<vector<double>> A(size, vector<double>(size, 1.5));
    vector<vector<double>> B(size, vector<double>(size, 2.5));
    vector<vector<double>> C(size, vector<double>(size, 0.0));

    for (int i = 0; i < size; ++i)
        for (int j = 0; j < size; ++j)
            for (int k = 0; k < size; ++k)
                C[i][j] += A[i][k] * B[k][j];

    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<seconds>(end - start).count();
    cout << "Matrix multiplication of size " << size << " completed (took " << elapsed << " seconds)" << endl;
}

void large_vector_sum(size_t size) {
    auto start = high_resolution_clock::now();

    vector<double> data(size);
    mt19937 rng(42);
    uniform_real_distribution<> dist(0.0, 1.0);

    for (auto& val : data)
        val = dist(rng);

    double sum = 0.0;
    for (const auto& val : data)
        sum += val;

    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<seconds>(end - start).count();
    cout << "Sum of large vector: " << sum << " (took " << elapsed << " seconds)" << endl;
}

int main() {
    auto total_start = high_resolution_clock::now();

    calculate_primes(10'000'000);
    matrix_multiply(500);
    large_vector_sum(100'000'000);

    auto total_end = high_resolution_clock::now();
    auto total_duration = duration_cast<seconds>(total_end - total_start).count();
    cout << "Total time taken by all tasks: " << total_duration << " seconds" << endl;

    return 0;
}
```

Here are the results

```log
C:\Apptest>.\app_arm64.exe
Total primes up to 10000000: 664579 (took 6 seconds)
Matrix multiplication of size 500 completed (took 2 seconds)
Sum of large vector: 5.00028e+07 (took 6 seconds)
Total time taken by all tasks: 14 seconds

C:\Apptest>.\app_arm64ec.exe
Total primes up to 10000000: 664579 (took 6 seconds)
Matrix multiplication of size 500 completed (took 1 seconds)
Sum of large vector: 5.00028e+07 (took 6 seconds)
Total time taken by all tasks: 14 seconds

C:\Apptest>.\app_x64.exe
Total primes up to 10000000: 664579 (took 14 seconds)
Matrix multiplication of size 500 completed (took 3 seconds)
Sum of large vector: 5.00028e+07 (took 5 seconds)
Total time taken by all tasks: 23 seconds
```
